src/main/java/org/spongepowered/api/command/spec/CommandSpec.java

```diff
@@ -24,439 +24,202 @@
+import org.spongepowered.api.command.parameters.flags.Flags;
+import org.spongepowered.api.command.parameters.tokens.InputTokenizer;
 import org.spongepowered.api.text.Text;
-import org.spongepowered.api.world.Location;
-import org.spongepowered.api.world.World;
+import org.spongepowered.api.util.ResettableBuilder;
 
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.List;
+import javax.annotation.Nullable;
 import java.util.Map;
 import java.util.Optional;
 
-import javax.annotation.Nullable;
-
 /**
- * Specification for how command arguments should be parsed.
+ * A high level interface for {@link CommandCallable}s that handles argument
+ * parsing and subcommand handling.
  */
-public final class CommandSpec implements CommandCallable {
-
-    private final CommandElement args;
-    private final CommandExecutor executor;
-    private final Optional<Text> description;
-    private final Optional<Text> extendedDescription;
-    @Nullable private final String permission;
-    private final InputTokenizer argumentParser;
-
-    CommandSpec(CommandElement args, CommandExecutor executor, @Nullable Text description, @Nullable Text extendedDescription,
-            @Nullable String permission, InputTokenizer parser) {
-        this.args = args;
-        this.executor = executor;
-        this.permission = permission;
-        this.description = Optional.ofNullable(description);
-        this.extendedDescription = Optional.ofNullable(extendedDescription);
-        this.argumentParser = parser;
-    }
+public interface CommandSpec extends CommandCallable {
 
     /**
-     * Return a new builder for a CommandSpec.
+     * Gets a builder for building a {@link CommandSpec}
      *
-     * @return a new builder
+     * @return The {@link Builder}
      */
-    public static Builder builder() {
-        return new Builder();
+    static Builder builder() {
+        return Sponge.getRegistry().createBuilder(Builder.class);
     }
 
     /**
-     * Builder for command specs.
+     * Return a longer description for this command. This description is
+     * composed of at least all present of the short description, the usage
+     * statement, and the extended description
+     *
+     * @param source The source to get the extended description for
+     * @return the extended description
      */
-    public static final class Builder {
-
-        private static final CommandElement DEFAULT_ARG = GenericArguments.none();
-        private CommandElement args = DEFAULT_ARG;
-        @Nullable
-        private Text description;
-        @Nullable
-        private Text extendedDescription;
-        @Nullable
-        private String permission;
-        @Nullable
-        private CommandExecutor executor;
-        @Nullable
-        private Map<List<String>, CommandCallable> childCommandMap;
-        private InputTokenizer argumentParser = InputTokenizer.quotedStrings(false);
+    Optional<Text> getExtendedDescription(CommandSource source);
```

Comment:
I don't think you have nearly enough getters to be useful for CommandSpec.

I'd like to see methods of getting access to the children, their aliases, permissions, and help, in order to assist help plugins listing sub commands.

Granted all this becomes moot if a plugin implements a dispatcher manually, so maybe these should / could be inherited from an interface, that plugins could potentially implement to help out.